From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Rezoss <rezoss@unica>
Date: Wed, 22 Jul 2026 22:00:00 +0200
Subject: [PATCH] Restore CHN Now Nudge conversation bridge

Recover chat compose nodes, synthesize chat user context, add fallback
app-view metadata, debounce conversation parsing on relevant events, and
bootstrap SmartReplyHandler so SmartReplyResolver is reached.
---
diff --git a/smali_classes14/com/samsung/android/smartsuggestions/extservice/services/ContentCaptureService.smali b/smali_classes14/com/samsung/android/smartsuggestions/extservice/services/ContentCaptureService.smali
--- a/smali_classes14/com/samsung/android/smartsuggestions/extservice/services/ContentCaptureService.smali
+++ b/smali_classes14/com/samsung/android/smartsuggestions/extservice/services/ContentCaptureService.smali
@@ -413,6 +413,18 @@
 
     if-eqz v0, :cond_rezoss_telegram_packages_done
 
+    const-string v1, "com.whatsapp"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.whatsapp.w4b"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    const-string v1, "com.samsung.android.messaging"
+
+    invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
     const-string/jumbo v1, "org.telegram.messenger"
 
     invoke-interface {v0, v1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
diff --git a/smali_classes15/com/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge.smali b/smali_classes15/com/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge.smali
new file mode 100644
--- /dev/null
+++ b/smali_classes15/com/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge.smali
@@ -0,0 +1,616 @@
+.class public final Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;
+.super Ljava/lang/Object;
+.source "SourceFile"
+
+
+# direct methods
+.method private constructor <init>()V
+    .locals 0
+
+    invoke-direct {p0}, Ljava/lang/Object;-><init>()V
+
+    return-void
+.end method
+
+.method private static final add(Ljava/util/Set;Ljava/lang/String;)V
+    .locals 0
+
+    invoke-interface {p0, p1}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    return-void
+.end method
+
+.method private static final containsToken(Ljava/lang/String;Ljava/lang/String;)Z
+    .locals 1
+
+    if-eqz p0, :cond_false
+
+    invoke-static {p0, p1}, Lbl0/c;->h(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v0
+
+    if-nez v0, :cond_true
+
+    invoke-static {p0, p1}, Lkotlin/text/v;->l0(Ljava/lang/String;Ljava/lang/String;)Z
+
+    move-result v0
+
+    if-eqz v0, :cond_false
+
+    :cond_true
+    const/4 p0, 0x1
+
+    return p0
+
+    :cond_false
+    const/4 p0, 0x0
+
+    return p0
+.end method
+
+.method public static final findComposeNode(Landroid/app/assist/AssistStructure$ViewNode;)Landroid/app/assist/AssistStructure$ViewNode;
+    .locals 4
+
+    if-eqz p0, :cond_not_found
+
+    invoke-static {p0}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->isComposeNode(Landroid/app/assist/AssistStructure$ViewNode;)Z
+
+    move-result v0
+
+    if-eqz v0, :cond_children
+
+    return-object p0
+
+    :cond_children
+    invoke-virtual {p0}, Landroid/app/assist/AssistStructure$ViewNode;->getChildCount()I
+
+    move-result v0
+
+    const/4 v1, 0x0
+
+    :goto_children
+    if-ge v1, v0, :cond_not_found
+
+    invoke-virtual {p0, v1}, Landroid/app/assist/AssistStructure$ViewNode;->getChildAt(I)Landroid/app/assist/AssistStructure$ViewNode;
+
+    move-result-object v2
+
+    invoke-static {v2}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->findComposeNode(Landroid/app/assist/AssistStructure$ViewNode;)Landroid/app/assist/AssistStructure$ViewNode;
+
+    move-result-object v3
+
+    if-eqz v3, :cond_next
+
+    return-object v3
+
+    :cond_next
+    add-int/lit8 v1, v1, 0x1
+
+    goto :goto_children
+
+    :cond_not_found
+    const/4 p0, 0x0
+
+    return-object p0
+.end method
+
+.method public static final isComposeNode(Landroid/app/assist/AssistStructure$ViewNode;)Z
+    .locals 3
+
+    if-eqz p0, :cond_false
+
+    invoke-virtual {p0}, Landroid/app/assist/AssistStructure$ViewNode;->getAutofillType()I
+
+    move-result v0
+
+    const/4 v1, 0x1
+
+    if-ne v0, v1, :cond_false
+
+    invoke-virtual {p0}, Landroid/app/assist/AssistStructure$ViewNode;->getIdEntry()Ljava/lang/String;
+
+    move-result-object v0
+
+    const-string v2, "entry"
+
+    invoke-static {v0, v2}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->containsToken(Ljava/lang/String;Ljava/lang/String;)Z
+
+    move-result v2
+
+    if-nez v2, :cond_true
+
+    const-string v2, "compose"
+
+    invoke-static {v0, v2}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->containsToken(Ljava/lang/String;Ljava/lang/String;)Z
+
+    move-result v2
+
+    if-nez v2, :cond_true
+
+    const-string v2, "input"
+
+    invoke-static {v0, v2}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->containsToken(Ljava/lang/String;Ljava/lang/String;)Z
+
+    move-result v2
+
+    if-nez v2, :cond_true
+
+    const-string v2, "edit"
+
+    invoke-static {v0, v2}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->containsToken(Ljava/lang/String;Ljava/lang/String;)Z
+
+    move-result v2
+
+    if-nez v2, :cond_true
+
+    invoke-virtual {p0}, Landroid/app/assist/AssistStructure$ViewNode;->getTextIdEntry()Ljava/lang/String;
+
+    move-result-object v0
+
+    const-string v2, "entry"
+
+    invoke-static {v0, v2}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->containsToken(Ljava/lang/String;Ljava/lang/String;)Z
+
+    move-result v2
+
+    if-nez v2, :cond_true
+
+    const-string v2, "compose"
+
+    invoke-static {v0, v2}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->containsToken(Ljava/lang/String;Ljava/lang/String;)Z
+
+    move-result v2
+
+    if-nez v2, :cond_true
+
+    const-string v2, "input"
+
+    invoke-static {v0, v2}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->containsToken(Ljava/lang/String;Ljava/lang/String;)Z
+
+    move-result v2
+
+    if-nez v2, :cond_true
+
+    const-string v2, "edit"
+
+    invoke-static {v0, v2}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->containsToken(Ljava/lang/String;Ljava/lang/String;)Z
+
+    move-result v2
+
+    if-nez v2, :cond_true
+
+    invoke-virtual {p0}, Landroid/app/assist/AssistStructure$ViewNode;->getHint()Ljava/lang/String;
+
+    move-result-object v0
+
+    const-string v2, "message"
+
+    invoke-static {v0, v2}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->containsToken(Ljava/lang/String;Ljava/lang/String;)Z
+
+    move-result v2
+
+    if-nez v2, :cond_true
+
+    const-string/jumbo v2, "reply"
+
+    invoke-static {v0, v2}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->containsToken(Ljava/lang/String;Ljava/lang/String;)Z
+
+    move-result v0
+
+    if-nez v0, :cond_true
+
+    invoke-virtual {p0}, Landroid/app/assist/AssistStructure$ViewNode;->getClassName()Ljava/lang/String;
+
+    move-result-object p0
+
+    const-string v0, "EditText"
+
+    invoke-static {p0, v0}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->containsToken(Ljava/lang/String;Ljava/lang/String;)Z
+
+    move-result v0
+
+    if-nez v0, :cond_true
+
+    const-string/jumbo v0, "editable"
+
+    invoke-static {p0, v0}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->containsToken(Ljava/lang/String;Ljava/lang/String;)Z
+
+    move-result p0
+
+    if-eqz p0, :cond_false
+
+    :cond_true
+    return v1
+
+    :cond_false
+    const/4 p0, 0x0
+
+    return p0
+.end method
+
+.method public static final isScheduleEvent(Landroid/view/contentcapture/ContentCaptureEvent;)Z
+    .locals 2
+
+    if-eqz p0, :cond_false
+
+    invoke-virtual {p0}, Landroid/view/contentcapture/ContentCaptureEvent;->getType()I
+
+    move-result p0
+
+    const/4 v0, 0x1
+
+    if-eq p0, v0, :cond_true
+
+    const/4 v1, 0x3
+
+    if-eq p0, v1, :cond_true
+
+    const/4 v1, 0x5
+
+    if-ne p0, v1, :cond_false
+
+    :cond_true
+    return v0
+
+    :cond_false
+    const/4 p0, 0x0
+
+    return p0
+.end method
+
+.method public static final isSupportedComponent(Landroid/content/ComponentName;)Z
+    .locals 0
+
+    if-eqz p0, :cond_false
+
+    invoke-virtual {p0}, Landroid/content/ComponentName;->getPackageName()Ljava/lang/String;
+
+    move-result-object p0
+
+    invoke-static {p0}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->isSupportedPackage(Ljava/lang/String;)Z
+
+    move-result p0
+
+    return p0
+
+    :cond_false
+    const/4 p0, 0x0
+
+    return p0
+.end method
+
+.method public static final isSupportedPackage(Ljava/lang/String;)Z
+    .locals 2
+
+    if-eqz p0, :cond_false
+
+    const-string v0, "com.whatsapp"
+
+    invoke-static {p0, v0}, Lbl0/c;->h(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_true
+
+    const-string v0, "com.whatsapp.w4b"
+
+    invoke-static {p0, v0}, Lbl0/c;->h(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_true
+
+    const-string v0, "com.samsung.android.messaging"
+
+    invoke-static {p0, v0}, Lbl0/c;->h(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_true
+
+    const-string/jumbo v0, "org.telegram.messenger"
+
+    invoke-static {p0, v0}, Lbl0/c;->h(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_true
+
+    const-string/jumbo v0, "org.telegram.messenger.web"
+
+    invoke-static {p0, v0}, Lbl0/c;->h(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_true
+
+    const-string/jumbo v0, "org.telegram.plus"
+
+    invoke-static {p0, v0}, Lbl0/c;->h(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_true
+
+    const-string/jumbo v0, "org.thunderdog.challegram"
+
+    invoke-static {p0, v0}, Lbl0/c;->h(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result p0
+
+    if-eqz p0, :cond_false
+
+    :cond_true
+    const/4 p0, 0x1
+
+    return p0
+
+    :cond_false
+    const/4 p0, 0x0
+
+    return p0
+.end method
+
+.method public static final populateAppViewFallback(Ljava/lang/String;Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/AppViewParsingSpec;)Z
+    .locals 3
+
+    invoke-static {p0}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->isSupportedPackage(Ljava/lang/String;)Z
+
+    move-result p0
+
+    if-eqz p0, :cond_false
+
+    invoke-virtual {p1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/AppViewParsingSpec;->getMessageIdEntries()Ljava/util/Set;
+
+    move-result-object v0
+
+    const-string/jumbo v1, "message_text"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string v1, "caption"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string/jumbo v1, "quoted_message_text"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string v1, "bubble_text_view"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string v1, "bubble_sent_text_preview"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string v1, "bubble_receive_text_preview"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string v1, "bubble_header_message"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string v1, "content_text_view"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string v1, "message"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string/jumbo v1, "text"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string/jumbo v1, "text1"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string/jumbo v1, "text2"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string v1, "body"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string v1, "content"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    invoke-virtual {p1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/AppViewParsingSpec;->getNonTextMessageIdEntries()Ljava/util/Set;
+
+    move-result-object v0
+
+    const-string v1, "bubble_image_view"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    invoke-virtual {p1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/AppViewParsingSpec;->getListviewIdEntries()Ljava/util/Set;
+
+    move-result-object v0
+
+    const-string v1, "list"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string v1, "conversation_list"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string v1, "bubble_list_view"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string/jumbo v1, "messages_list"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string v1, "chat_list"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string/jumbo v1, "recycler_view"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    invoke-virtual {p1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/AppViewParsingSpec;->getTitleIdEntries()Ljava/util/Set;
+
+    move-result-object v0
+
+    const-string v1, "conversation_contact_name"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string v1, "action_bar_title"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string/jumbo v1, "toolbar_title"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string v1, "extended_appbar_custom_title"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string v1, "composer_main_title_container"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string v1, "composer_title"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string/jumbo v1, "title"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string v1, "chat_title"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string v1, "conversation_title"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string v1, "name"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string v1, "profile_name"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    invoke-virtual {p1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/AppViewParsingSpec;->getAvatarIdEntries()Ljava/util/Set;
+
+    move-result-object v0
+
+    const-string v1, "conversation_contact_photo"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string v1, "bubble_avatar"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string/jumbo v1, "toolbar_avatar_image"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string/jumbo v1, "recipient_avatar"
+
+    invoke-static {v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    invoke-virtual {p1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/AppViewParsingSpec;->getConversationThreadingViewClasses()Ljava/util/Set;
+
+    move-result-object p0
+
+    const-string v0, "androidx.recyclerview.widget.RecyclerView"
+
+    invoke-static {p0, v0}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string v0, "android.widget.ListView"
+
+    invoke-static {p0, v0}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string v0, "android.support.v7.widget.RecyclerView"
+
+    invoke-static {p0, v0}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string v0, "com.samsung.android.messaging.ui.view.bubble.list.BubbleListView"
+
+    invoke-static {p0, v0}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string v0, "com.samsung.android.messaging.ui.view.widget.common.MessageRecyclerView"
+
+    invoke-static {p0, v0}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string p0, "ScreenContentManager"
+
+    const-string/jumbo v0, "Now Nudge conversation app-view fallback enabled"
+
+    invoke-static {p0, v0}, Landroid/util/Log;->i(Ljava/lang/String;Ljava/lang/String;)I
+
+    const/4 p0, 0x1
+
+    return p0
+
+    :cond_false
+    const/4 p0, 0x0
+
+    return p0
+.end method
+
+.method public static final synthesizeUserContext(Lcom/samsung/android/smartsuggestions/service/autofill/data/AssistData;Ljava/util/Map;)Z
+    .locals 3
+
+    if-eqz p0, :cond_false
+
+    if-eqz p1, :cond_false
+
+    invoke-virtual {p0}, Lcom/samsung/android/smartsuggestions/service/autofill/data/AssistData;->getComponentName()Landroid/content/ComponentName;
+
+    move-result-object v0
+
+    invoke-static {v0}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->isSupportedComponent(Landroid/content/ComponentName;)Z
+
+    move-result v0
+
+    if-eqz v0, :cond_false
+
+    invoke-virtual {p0}, Lcom/samsung/android/smartsuggestions/service/autofill/data/AssistData;->getNodeInfo()Landroid/app/assist/AssistStructure$ViewNode;
+
+    move-result-object v0
+
+    invoke-static {v0}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->isComposeNode(Landroid/app/assist/AssistStructure$ViewNode;)Z
+
+    move-result v0
+
+    if-eqz v0, :cond_false
+
+    invoke-virtual {p0}, Lcom/samsung/android/smartsuggestions/service/autofill/data/AssistData;->getFieldType()Lcom/samsung/android/smartsuggestions/service/autofill/data/FieldInputType;
+
+    move-result-object p0
+
+    const-string/jumbo v0, "message"
+
+    invoke-interface {p1, p0, v0}, Ljava/util/Map;->put(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;
+
+    const-string p0, "AutofillEventController"
+
+    const-string/jumbo v0, "synthesized chat text userContext"
+
+    invoke-static {p0, v0}, Lcom/samsung/android/smartsuggestions/service/autofill/util/AutofillFlowLogger;->i(Ljava/lang/String;Ljava/lang/String;)V
+
+    const/4 p0, 0x1
+
+    return p0
+
+    :cond_false
+    const/4 p0, 0x0
+
+    return p0
+.end method
diff --git a/smali_classes15/com/samsung/android/smartsuggestions/service/inlinesuggestions/data/AssistStructureFinder.smali b/smali_classes15/com/samsung/android/smartsuggestions/service/inlinesuggestions/data/AssistStructureFinder.smali
--- a/smali_classes15/com/samsung/android/smartsuggestions/service/inlinesuggestions/data/AssistStructureFinder.smali
+++ b/smali_classes15/com/samsung/android/smartsuggestions/service/inlinesuggestions/data/AssistStructureFinder.smali
@@ -229,7 +229,7 @@
 
 # virtual methods
 .method public final find(Landroid/content/ComponentName;Ljava/lang/String;)Landroid/app/assist/AssistStructure$ViewNode;
-    .locals 5
+    .locals 7
     .param p1    # Landroid/content/ComponentName;
         .annotation build Lorg/jetbrains/annotations/NotNull;
         .end annotation
@@ -249,6 +249,10 @@
 
     invoke-static {p2, v0}, Lbl0/c;->m(Ljava/lang/Object;Ljava/lang/String;)V
 
+    invoke-static {p1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->isSupportedComponent(Landroid/content/ComponentName;)Z
+
+    move-result v5
+
     invoke-static {}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/assist/AssistHolder;->getInstance()Lcom/samsung/android/smartsuggestions/service/deepscreencapture/assist/AssistHolder;
 
     move-result-object v0
@@ -290,6 +294,8 @@
 
     move-result-object v2
 
+    move-object v6, v2
+
     new-instance v3, Ljava/lang/StringBuilder;
 
     const-string/jumbo v4, "find: "
@@ -327,6 +333,21 @@
     return-object v2
 
     :cond_1
+    if-eqz v5, :cond_rezoss_next_window
+
+    invoke-static {v6}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->findComposeNode(Landroid/app/assist/AssistStructure$ViewNode;)Landroid/app/assist/AssistStructure$ViewNode;
+
+    move-result-object v2
+
+    if-eqz v2, :cond_rezoss_next_window
+
+    const-string p0, "Recovered messaging compose node for collectAssistData"
+
+    invoke-static {v4, p0}, Lxc/e;->c(Ljava/lang/String;Ljava/lang/String;)V
+
+    return-object v2
+
+    :cond_rezoss_next_window
     add-int/lit8 v1, v1, 0x1
 
     goto :goto_1
diff --git a/smali_classes15/com/samsung/android/smartsuggestions/service/autofill/collector/AutofillEventController.smali b/smali_classes15/com/samsung/android/smartsuggestions/service/autofill/collector/AutofillEventController.smali
--- a/smali_classes15/com/samsung/android/smartsuggestions/service/autofill/collector/AutofillEventController.smali
+++ b/smali_classes15/com/samsung/android/smartsuggestions/service/autofill/collector/AutofillEventController.smali
@@ -583,7 +583,7 @@
 .end method
 
 .method private final existToAddInfoToFields(Lcom/samsung/android/smartsuggestions/service/autofill/data/AssistData;Ljava/util/Map;)Z
-    .locals 1
+    .locals 2
     .annotation system Ldalvik/annotation/Signature;
         value = {
             "(",
@@ -595,6 +595,8 @@
         }
     .end annotation
 
+    move-object v1, p2
+
     invoke-virtual {p1}, Lcom/samsung/android/smartsuggestions/service/autofill/data/AssistData;->getFieldType()Lcom/samsung/android/smartsuggestions/service/autofill/data/FieldInputType;
 
     move-result-object p0
@@ -642,6 +644,17 @@
 
     :cond_1
     :goto_0
+    invoke-static {p1, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->synthesizeUserContext(Lcom/samsung/android/smartsuggestions/service/autofill/data/AssistData;Ljava/util/Map;)Z
+
+    move-result p0
+
+    if-eqz p0, :cond_rezoss_no_context
+
+    const/4 p0, 0x1
+
+    return p0
+
+    :cond_rezoss_no_context
     const-string p0, "There is no userContext for detected input field"
 
     invoke-static {p2, p0}, Lcom/samsung/android/smartsuggestions/service/autofill/util/AutofillFlowLogger;->i(Ljava/lang/String;Ljava/lang/String;)V
diff --git a/smali_classes15/com/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/ScreenContentManager.smali b/smali_classes15/com/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/ScreenContentManager.smali
--- a/smali_classes15/com/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/ScreenContentManager.smali
+++ b/smali_classes15/com/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/ScreenContentManager.smali
@@ -848,6 +848,19 @@
 
     invoke-static {v13, v14}, Lbl0/c;->l(Ljava/lang/Object;Ljava/lang/String;)V
 
+    move-object/from16 v14, p3
+
+    invoke-static {v13, v14}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->populateAppViewFallback(Ljava/lang/String;Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/AppViewParsingSpec;)Z
+
+    move-result v14
+
+    if-eqz v14, :cond_rezoss_appview_fallback_done
+
+    const/4 v14, 0x1
+
+    return v14
+
+    :cond_rezoss_appview_fallback_done
     iget-object v14, v0, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/ScreenContentManager;->appViewParsingSpecJsonString:Ljava/util/HashMap;
 
     invoke-virtual {v14, v1}, Ljava/util/HashMap;->get(Ljava/lang/Object;)Ljava/lang/Object;
@@ -2638,13 +2651,11 @@
     if-eqz v1, :cond_20
 
     :cond_1f
-    invoke-virtual {v6}, Landroid/view/contentcapture/ContentCaptureEvent;->getType()I
+    invoke-static {v6}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->isScheduleEvent(Landroid/view/contentcapture/ContentCaptureEvent;)Z
 
     move-result v1
 
-    const/4 v3, 0x5
-
-    if-ne v1, v3, :cond_20
+    if-eqz v1, :cond_20
 
     invoke-direct {v0, v2}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/ScreenContentManager;->makeScheduleForLiveTranslationBatchProcessing(Landroid/view/contentcapture/ContentCaptureSessionId;)V
 
@@ -2899,7 +2910,7 @@
 
 # virtual methods
 .method public final addLiveTranslationObserver(Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/ChatTranslationObserver;)V
-    .locals 3
+    .locals 2
     .param p1    # Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/ChatTranslationObserver;
         .annotation build Lorg/jetbrains/annotations/NotNull;
         .end annotation
@@ -2909,19 +2920,25 @@
 
     invoke-static {p1, v0}, Lbl0/c;->m(Ljava/lang/Object;Ljava/lang/String;)V
 
-    sget-object v0, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/ExecutorHolder;->INSTANCE:Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/ExecutorHolder;
-
-    invoke-virtual {v0}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/ExecutorHolder;->getCcsExecutor()Ljava/util/concurrent/ScheduledExecutorService;
-
-    move-result-object v0
-
-    new-instance v1, Lcom/samsung/android/smartsuggestions/feature/aisuggestion/ui/view/cover/j;
-
-    const/16 v2, 0x9
-
-    invoke-direct {v1, p0, p1, v2}, Lcom/samsung/android/smartsuggestions/feature/aisuggestion/ui/view/cover/j;-><init>(Ljava/lang/Object;Ljava/lang/Object;I)V
-
-    invoke-interface {v0, v1}, Ljava/util/concurrent/Executor;->execute(Ljava/lang/Runnable;)V
+    iget-object v0, p0, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/ScreenContentManager;->chatTranslationObserver:Ljava/util/concurrent/CopyOnWriteArrayList;
+
+    invoke-virtual {v0, p1}, Ljava/util/concurrent/CopyOnWriteArrayList;->contains(Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_rezoss_observer_registered
+
+    iget-object p0, p0, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/ScreenContentManager;->chatTranslationObserver:Ljava/util/concurrent/CopyOnWriteArrayList;
+
+    invoke-virtual {p0, p1}, Ljava/util/concurrent/CopyOnWriteArrayList;->add(Ljava/lang/Object;)Z
+
+    const-string p0, "ScreenContentManager"
+
+    const-string/jumbo p1, "registered chat observer synchronously"
+
+    invoke-static {p0, p1}, Landroid/util/Log;->i(Ljava/lang/String;Ljava/lang/String;)I
+
+    :cond_rezoss_observer_registered
 
     return-void
 .end method
diff --git a/smali_classes15/com/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler.smali b/smali_classes15/com/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler.smali
--- a/smali_classes15/com/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler.smali
+++ b/smali_classes15/com/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler.smali
@@ -1414,7 +1414,19 @@
 .end method
 
 .method private final isEnabledPackage(Landroid/content/Context;Landroid/content/ComponentName;)Z
-    .locals 0
+    .locals 1
+
+    invoke-static {p2}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->isSupportedComponent(Landroid/content/ComponentName;)Z
+
+    move-result p0
+
+    if-eqz p0, :cond_rezoss_original_enabled_package
+
+    const/4 p0, 0x1
+
+    return p0
+
+    :cond_rezoss_original_enabled_package
 
     invoke-static {p1}, Lcom/samsung/android/smartsuggestions/service/policy/PolicyManager;->getInstance(Landroid/content/Context;)Lcom/samsung/android/smartsuggestions/service/policy/PolicyManager;
 
@@ -1775,6 +1787,25 @@
     :goto_0
     const-string v8, "SmartReplyHandler[reply]"
 
+    invoke-static {v2}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->isSupportedPackage(Ljava/lang/String;)Z
+
+    move-result v5
+
+    if-eqz v5, :cond_rezoss_start_original_enabled
+
+    const/4 v1, 0x1
+
+    sput-boolean v1, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler;->isNowNudgeOn:Z
+
+    sput-boolean v1, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler;->isKeyboardVisible:Z
+
+    const-string v5, "force-enabled chat startSmartReply gate"
+
+    invoke-static {v8, v5}, Landroid/util/Log;->i(Ljava/lang/String;Ljava/lang/String;)I
+
+    goto :cond_1
+
+    :cond_rezoss_start_original_enabled
     if-nez v1, :cond_1
 
     const-string/jumbo v0, "startSmartReply, enabled: false"
@@ -2082,7 +2113,7 @@
 .end method
 
 .method public onChatTranslationMessages(Landroid/view/contentcapture/ContentCaptureSessionId;Landroid/content/Context;Ljava/util/List;Landroid/content/ComponentName;Landroid/graphics/Rect;)V
-    .locals 2
+    .locals 3
     .param p1    # Landroid/view/contentcapture/ContentCaptureSessionId;
         .annotation build Lorg/jetbrains/annotations/NotNull;
         .end annotation
@@ -2133,6 +2164,77 @@
 
     invoke-static {p4, v0}, Lbl0/c;->m(Ljava/lang/Object;Ljava/lang/String;)V
 
+    invoke-static {p4}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->isSupportedComponent(Landroid/content/ComponentName;)Z
+
+    move-result v0
+
+    if-eqz v0, :cond_rezoss_bootstrap_done
+
+    sget-object v0, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler;->componentNameMap:Ljava/util/HashMap;
+
+    invoke-virtual {v0, p1, p4}, Ljava/util/HashMap;->put(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;
+
+    invoke-virtual {p4}, Landroid/content/ComponentName;->getPackageName()Ljava/lang/String;
+
+    move-result-object v0
+
+    sget-object v1, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler;->isPackageEnabledMap:Ljava/util/HashMap;
+
+    sget-object v2, Ljava/lang/Boolean;->TRUE:Ljava/lang/Boolean;
+
+    invoke-interface {v1, v0, v2}, Ljava/util/Map;->put(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;
+
+    sget-object v0, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler;->newMessagesSessionMap:Ljava/util/HashMap;
+
+    invoke-virtual {v0, p1}, Ljava/util/HashMap;->get(Ljava/lang/Object;)Ljava/lang/Object;
+
+    move-result-object v1
+
+    if-eqz v1, :cond_rezoss_create_new_messages
+
+    sget-object v0, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler;->pastMessagesSessionMap:Ljava/util/HashMap;
+
+    goto :goto_rezoss_past_messages
+
+    :cond_rezoss_create_new_messages
+    new-instance v1, Ljava/util/ArrayList;
+
+    invoke-direct {v1}, Ljava/util/ArrayList;-><init>()V
+
+    invoke-virtual {v0, p1, v1}, Ljava/util/HashMap;->put(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;
+
+    sget-object v0, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler;->pastMessagesSessionMap:Ljava/util/HashMap;
+
+    :goto_rezoss_past_messages
+    invoke-virtual {v0, p1}, Ljava/util/HashMap;->get(Ljava/lang/Object;)Ljava/lang/Object;
+
+    move-result-object v1
+
+    if-eqz v1, :cond_rezoss_create_past_messages
+
+    goto :goto_rezoss_bootstrap_state
+
+    :cond_rezoss_create_past_messages
+    new-instance v1, Ljava/util/ArrayList;
+
+    invoke-direct {v1}, Ljava/util/ArrayList;-><init>()V
+
+    invoke-virtual {v0, p1, v1}, Ljava/util/HashMap;->put(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;
+
+    :goto_rezoss_bootstrap_state
+    const/4 v0, 0x1
+
+    sput-boolean v0, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler;->isNowNudgeOn:Z
+
+    sput-boolean v0, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler;->isKeyboardVisible:Z
+
+    const-string v0, "SmartReplyHandler[reply]"
+
+    const-string/jumbo v1, "bootstrapped chat SmartReply session"
+
+    invoke-static {v0, v1}, Landroid/util/Log;->i(Ljava/lang/String;Ljava/lang/String;)I
+
+    :cond_rezoss_bootstrap_done
     new-instance v0, Ljava/lang/StringBuilder;
 
     const-string/jumbo v1, "onChatTranslationMessages, sessionId="
@@ -2188,6 +2290,33 @@
 
     invoke-direct {p0, p2, p3, p1, p5}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler;->handleMessages(Landroid/content/Context;Ljava/util/List;Landroid/view/contentcapture/ContentCaptureSessionId;Landroid/graphics/Rect;)V
 
+    invoke-static {p4}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->isSupportedPackage(Ljava/lang/String;)Z
+
+    move-result p3
+
+    if-eqz p3, :cond_rezoss_dispatch_done
+
+    sget-object p3, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler;->replySentMap:Ljava/util/HashMap;
+
+    invoke-virtual {p3, p1}, Ljava/util/HashMap;->remove(Ljava/lang/Object;)Ljava/lang/Object;
+
+    const/4 p3, 0x1
+
+    sput-boolean p3, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler;->isNowNudgeOn:Z
+
+    sput-boolean p3, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler;->isKeyboardVisible:Z
+
+    const-string p3, "SmartReplyHandler[reply]"
+
+    const-string/jumbo p5, "force-dispatched chat SmartReply from parsed messages"
+
+    invoke-static {p3, p5}, Landroid/util/Log;->i(Ljava/lang/String;Ljava/lang/String;)I
+
+    invoke-direct {p0, p2, p1, p4}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler;->startSmartReply(Landroid/content/Context;Landroid/view/contentcapture/ContentCaptureSessionId;Ljava/lang/String;)V
+
+    return-void
+
+    :cond_rezoss_dispatch_done
     sget-boolean p3, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/SmartReplyHandler;->isKeyboardVisible:Z
 
     if-nez p3, :cond_1
